db/keyval/filesystem/client.go

```
@@ -0,0 +1,355 @@
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ligato/cn-infra/datasync"
+	"github.com/ligato/cn-infra/db/keyval"
+	"github.com/ligato/cn-infra/logging"
+)
+
+// Client arranges communication between file system and internal database, and is also responsible for upstream events.
+type Client struct {
+	sync.Mutex
+	log logging.Logger
+
+	// A list of filesystem paths. It may be specific file, or the whole directory. In case a path is a directory,
+	// all files within are processed.
+	paths []string
+
+	// Internal database mirrors changes in file system. Since the configuration can be only read, it is up to client
+	// to handle difference between configuration revisions. Evert database entry consists from three values; path
+	// of a file where the configuration is written, data key and data value.
+	// Note: database holds only configuration intended for agent with defined prefix
+	db FilesSystemDB
+
+	// File system reader, grants access to methods needed for checking/reading of system files
+	r reader.API
+
+	// Watcher watches over events incoming from the registered files.
+	watcher  *fsnotify.Watcher
+	watchers map[string]chan keyedData
+
+	// Prefix to recognize configuration for specific agent instance
+	agentPrefix string
+}
+
+// NewClient initializes file watcher, database and registers paths provided via plugin configuration file
+func NewClient(paths []string, prefix string, rd reader.API, log logging.Logger) (*Client, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to init file watcher: %v", err)
+	}
+
+	// Init client object
+	c := &Client{
+		paths:       paths,
+		watcher:     watcher,
+		watchers:    make(map[string]chan keyedData),
+		db:          NewDbClient(),
+		r:           rd,
+		agentPrefix: prefix,
+		log:         log,
+	}
+
+	// Do the initial read and store everything to internal database
+	for _, path := range paths {
+		// Register to watcher
+		watcher.Add(path)
```

Error is being ignored here.